### ISSUE TITLE
No data propogated to Materialized View if column names mismatch

### ISSUE THREAD

**Issue by tavplubix** [ClickHouse/ClickHouse#59072]
```
CREATE TABLE default.src
(
    `n` Int32
)
ENGINE = Log;
CREATE TABLE default.t
(
    `n` Int32
)
ENGINE = Log;
CREATE TABLE default.url
(
    `s` String
)
ENGINE = URL('http://localhost:8123/?query=insert+into+t+select+0+where+0', 'TSVRaw');
CREATE MATERIALIZED VIEW default.mv TO default.url
(
    `query` String
) AS
SELECT concat(' union all select ', toString(n)) as query
FROM default.src;

insert into src values (123);
insert into url values (' union all select 456');

select * from t; -- this returns empty result, not was inserted
```

Last time it worked in 19.10: https://fiddle.clickhouse.com/84b9ac3c-b610-4743-9a0c-e44b28b721ff
Theh since 19.11 it throws an exception: https://fiddle.clickhouse.com/697e16c2-bb18-401c-8439-589746874eef
And then since 19.16 it silently ignores data and inserts nothing: https://fiddle.clickhouse.com/a9a25a1b-2e99-498b-8331-803913914fd8

And this works: https://fiddle.clickhouse.com/b8a97514-ed88-43ff-aee2-095d1cf04a1a
The difference is the column name. MV only inserts columns if their names exactly match column names in the destination table. It's very error-prone because you can miss some data due to a typo in a column name.

At least, we should throw an exception if 0 columns match.

**Comment by UnamedRus** [comment]
```
Received exception from server (version 19.11.14):
Code: 8. DB::Exception: Received from localhost:9000. DB::Exception: Cannot find column `s` in source stream: while pushing to view default.mv. 
```

Exact form of that exception isn't nice, what if you don't want to populate all columns in destination table from that particular MV.

**Comment by tavplubix** [comment]
Yes, I understand why it happens, but what do you think about "we should throw an exception if 0 columns match"? Or maybe log a warning...

**Comment by den-crane** [comment]
>Or maybe log a warning...

BTW, other databases have Query Warnings and JDBC has something like `getWarnings()`.
I would extend https://github.com/ClickHouse/ClickHouse/pull/58929 and return warnings using HTTP headers and native protocol.

**Comment by Algunenano** [comment]
> although, I have no idea how it works with HTTP when a query has resultset to send

It's not supported, because there isn't a standarized way of doing it in HTTP1. Long time ago I made this: https://github.com/ClickHouse/ClickHouse/pull/28204

**Comment by al13n321** [comment]
I'm adding a check for this on MV creation (along the way with refreshable MV work). CREATE MATERIALIZED VIEW will fail if the columns returned by the SELECT query are either missing from the target table or have incompatible type.

Question: when creating MV, should we require that all columns returned by the SELECT exist in the target table? Or that at least one of the columns exists? Guess I'll just put it behind a setting.

If some of the columns are later removed from the target table using ALTER, the MV will still work as long as at least one of the SELECTed columns remains, just like today.

(Refreshable MVs will be strict, both at CREATE time and at refresh time. I.e. some of the SELECTed columns are removed from the target table, refreshes will be failing. The user may then do ALTER MODIFY QUERY to make the SELECT query match the new table structure.)

**Comment by al13n321** [comment]
> At least, we should throw an exception if 0 columns match.

Done:
```
Received exception from server (version 25.1.5):
Code: 8. DB::Exception: Received from localhost:9000. DB::Exception: None of the columns produced by the SELECT query are present in the target table. Use 'AS' to assign aliases that match column names.. (THERE_IS_NO_COLUMN)
(query: CREATE MATERIALIZED VIEW default.mv TO default.url
(
    `query` String
) AS
SELECT concat(' union all select ', toString(n)) as query
FROM default.src;)
```
